Add adpd410x support for reading a large number of bytes from the FIFO at once

NSagan271 opened the pull request:
Add function to adpd410x.c to read a specified number of bytes from the
FIFO using the minimum number of register reads. Also add support for
reading a specified number of bytes (up to 255) from a register with one
i2c or spi read.

When reading large amounts of data from the FIFO (on the order of 100 - 512 bytes),
this greatly improves efficiency, as opposed to reading one packet from the FIFO
at a time.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1056) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1056) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=1056) it.</sub>

adrimbarean commented:
Try to limit the commit title to 75 characters.

adrimbarean commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
 
 	switch (dev->dev_type) {
 	case ADPD4100:
+		buff = (uint8_t *) calloc(num_bytes + 2, 1);
```
Would be better to have `buff = (uint8_t *) calloc(num_bytes + 2, sizeof(uint8_t));`
It is clearer this way.

adrimbarean commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
+	uint8_t i;
+	uint8_t *buff;
+
+	if (num_bytes > 255)
```
Is there a reason for this cap or is it arbitrary?

NSagan271 replied:
For I2C, reads are at most 255 bytes. SPI doesn't seem to have this limit, so I'll move the cap to the ADPD4101 section.

adrimbarean commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -551,22 +542,88 @@ static int32_t adpd410x_get_data_packet(struct adpd410x_dev *dev,
+		bytes_read += next_packet_size;
+	}
+
+	if (data == NULL)
```
Would it make more sense to check this at the beginning, before doing any reads?

adrimbarean reviewed (commented):
Looks good. A couple of items and I think we should be set.

NSagan271 commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
+{
+	int32_t ret;
+	uint8_t i;
+	uint8_t *buff;
```
I found out that, though I2C reads are at most 255 bytes, SPI reads can be longer (the bytes_number parameter for an SPI read is a 16-bit integer). I think in that case it might be better to dynamically allocate memory?

NSagan271 commented:
Updated PR, addressing all comments except [https://github.com/analogdevicesinc/no-OS/pull/1056#discussion_r671274019](https://github.com/analogdevicesinc/no-OS/pull/1056#discussion_r671274019).

adrimbarean reviewed (approved):
Looks good to me.

buha reviewed (approved):
looks good to me, any other opinions ?

mchindri commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
 	case ADPD4101:
+		// Number of bytes for an I2C read is an 8-bit number, or at most 255
+		num_bytes = min(num_bytes, 255);
+		return FAILURE;
```
this is not ok here, or should be inside a if.

mchindri commented on drivers/photo-electronic/adpd410x/adpd410x.c:
```diff
@@ -62,37 +62,74 @@ int32_t adpd410x_reg_read(struct adpd410x_dev *dev, uint16_t address,
 		break;
 	case ADPD4101:
+		// Number of bytes for an I2C read is an 8-bit number, or at most 255
+		num_bytes = min(num_bytes, 255);
```
I won't do this. Return failure if num_bytes > 255 or read all bytes inside a for.
A user will expect all values to be read.

NSagan271 commented:
Updated PR, addressing @mchindri's requested changes. I am returning FAILURE if num_bytes > 255 and the device type is ADPD4101.
